scripts/codemods/typed_flags.py

```diff
@@ -0,0 +1,182 @@
+        all_flag_classes = []
+        for attr_name in dir(flags):
+            obj = getattr(flags, attr_name)
+            if (
+                isinstance(obj, type)
+                and issubclass(obj, BASE_FLAG_CLASSES)
+                and obj not in BASE_FLAG_CLASSES
+            ):
+                all_flag_classes.append(obj.__name__)
+        self.flag_classes = all_flag_classes
+
+        return super().transform_module(tree)
+
+    def visit_ClassDef(self, node: cst.ClassDef) -> Optional[bool]:
+        # no reason to continue into classes
+        return False
+
+    def leave_ClassDef(self, _: cst.ClassDef, node: cst.ClassDef):
+        if not m.matches(node.name, m.OneOf(*map(m.Name, self.flag_classes))):
+            return node
+
+        # now the meat of the function
+        # we have two options here, and both of them depend on if __init__ exists, and *how* __init__ exists.
+        # thankfully, we aren't limited to just codemodding, we can import the module if we need it
+        # if __init__ exists without a TYPE_CHECKING block, we create and add two overloads
+        # if __init__ exists under a TYPE_CHECKING block, we remove and replace it
+        # we also check and remove previously existing TYPE_CHECKING and overloads
+
+        # and we have to import the module anyways, to get a list of valid flags for each subclass.
+        flag: flags.BaseFlags = getattr(flags, node.name.value)
+        all_flags = sorted(flag.VALID_FLAGS.keys())
+
+        # determine which method to use for creating the init overload, and possibly the update overload too.
+        # only a few of these need to be overloads, in which we can use the same technique from permission overloads.
+        # for the other ones, we need to find a TYPE_CHECKING block that *may* or may not contain a function definition.
+        # if none of the TYPE_CHECKING blocks contain a function definition, we make a new one and
+        # insert it near the beginning of the class body.
+        # we also decorate with @_generated so we can delete it later.
+
+        hide_behind_typechecking: bool = True
+        if_block: Optional[cst.If] = None
+        init: Optional[cst.FunctionDef] = None
+        body = list(node.body.body)
+        kwonly_params = [
+            cst.Param(cst.Name(flag_name), cst.Annotation(cst.Name("bool")), default=cst.Ellipsis())
+            for flag_name in all_flags
+        ]
+        for b in body.copy():
+            if m.matches(b, m.FunctionDef(m.Name("__init__"))) and isinstance(b, cst.FunctionDef):
+                deco_names = [getattr(deco.decorator, "value", "") for deco in b.decorators]
+                if "_generated" in deco_names:
+                    body.remove(b)
+                    continue
+                hide_behind_typechecking = False
+                init = b
+                break
+        node = node.with_deep_changes(node.body, body=body)
+
+        if hide_behind_typechecking:
```

Comment:
could replace this with a `if not init:` and get rid of the extra variable

Reply:
indirectly resolved with https://github.com/DisnakeDev/disnake/pull/778/commits/f00d250d2d811b048047ced7ae3a33a5a39537f8

Reply:
This still applies

Reply:
no it doesn't, init is defined at that point no matter if we used a typechecking hide or not.

Reply:
Ah, now I see. Thought `add_needed_import` worked differently.

Reply:
Changing this would still help avoid an assertion further down, though.

Reply:
https://github.com/DisnakeDev/disnake/pull/778/commits/2ec8c41a6f55de589bc7dd9472b0f5ac90a203a3